net/HttpRequest.hpp

```diff
@@ -1426,33 +1426,37 @@ class Session final : public ProtocolHandlerInterface
 
         assert(!!_response && "Response must be set!");
 
+        LOG_TRC("Starting syncRequest: " << _request.getVerb() << ' ' << host() << ':' << port()
+                                         << ' ' << _request.getUrl());
+
+        ConnectFailCallback origOnConnectFail = _onConnectFail;
+
         if (!isConnected())
         {
-            std::shared_ptr<StreamSocket> socket = connect();
-            if (!socket)
-            {
+            // Intercept onConnectFail to detect connection failure, call the original (if any),
+            // and finish the response to end the poller loop.
+            setConnectFailHandler([this, origOnConnectFail](const std::shared_ptr<http::Session>& rSession) {
                 LOG_ERR("Failed to connect to " << _host << ':' << _port);
-                return false;
-            }
-
-            poller.insertNewSocket(socket);
+                origOnConnectFail(rSession);
+                assert(_response->state() != Response::State::Complete && "unexpected Complete state");
+                _response->finish();
+            });
+            asyncConnect(poller);
         }
 
-        LOG_TRC("Starting syncRequest: " << _request.getVerb() << ' ' << host() << ':' << port()
-                                         << ' ' << _request.getUrl());
-
         poller.poll(timeout);
         while (!_response->done())
         {
             const auto now = std::chrono::steady_clock::now();
             if (checkTimeout(now))
-                return false;
+                break;
 
             const auto remaining =
                 std::chrono::duration_cast<std::chrono::microseconds>(deadline - now);
             poller.poll(remaining);
         }
 
+        setConnectFailHandler(origOnConnectFail);
```

Comment:
This isn't exception-safe. Yet another problem with the need to switch handlers, which we should avoid by having a more homogeneous design.